### ISSUE TITLE
java.lang.ClassCastException: com.google.common.collect.Maps$TransformedEntriesMap

### ISSUE THREAD

**Issue by j0hnsa** [cerberustesting/cerberus-source#1311]
Hello,

when using Automation name Uiautomator2, which is highly recommended to use with the latest versions of appium, this exception is thrown (Action Failed. java.lang.ClassCastException: com.google.common.collect.Maps$TransformedEntriesMap cannot be cast to java.lang.String) as soon as it try's to locate the 1st element:

http://192.168.134.17/Cerberus/ExecutionDetail2.jsp?executionId=46733131

On appium logs no errors are displayed:
,"android.intent.category.LAUNCHER","-f","0x10200000"]
[debug] [ADB] Waiting for activity matching pkg: 'com.ocito.laredoute' and activity: 'com.cstech.alpha.SplashScreen' to be focused
[debug] [ADB] Possible activities, to be checked: com.cstech.alpha.SplashScreen, com.ocito.laredoute.com.cstech.alpha.SplashScreen
[debug] [ADB] Getting focused package and activity
[debug] [ADB] Getting connected devices...
[debug] [ADB] 1 device(s) connected
[debug] [ADB] Running 'C:\Users\it140258\AppData\Local\Android\android-sdk\platform-tools\adb.exe' with args: ["-P",5037,"-s","3300ec51d0d3a269","shell","dumpsys","window","windows"]
[debug] [ADB] Found package: 'com.ocito.laredoute' and fully qualified activity name : 'com.cstech.alpha.SplashScreen'
[Appium] New AndroidUiautomator2Driver session created successfully, session a0ac2a7d-a5aa-428a-9044-c98d3061bbde added to master session list
[debug] [BaseDriver] Event 'newSessionStarted' logged at 1495640768403 (16:46:08 GMT+0100 (Hora de Verão de GMT))
[debug] [MJSONWP] Responding to client with driver.createSession() result: {"automationName":"Uiautomator2","platform":"ANDROID","app":"/LaredouteApps/app-preprod-8_0_227.apk","acceptSslCerts":true,"platformName":"Android","fullReset":true,"deviceName":"3300ec51d0d3a269"
,"clearSystemFiles":true,"udid":"3300ec51d0d3a269"}
[HTTP] <-- POST /wd/hub/session 200 33861 ms - 319
[HTTP] --> POST /wd/hub/session/a0ac2a7d-a5aa-428a-9044-c98d3061bbde/element {"using":"id","value":"com.ocito.laredoute:id/tv_selected_country"}
[debug] [MJSONWP] Calling AppiumDriver.findElement() with args: ["id","com.ocito.laredoute:id/tv_selected_country","a0ac2a7d-a5aa-428a-9044-c98d3061bbde"]
[debug] [BaseDriver] Valid locator strategies for this request: xpath, id, class name, accessibility id, -android uiautomator
[debug] [BaseDriver] Valid locator strategies for this request: xpath, id, class name, accessibility id, -android uiautomator
[debug] [BaseDriver] Waiting up to 0 ms for condition
[debug] [JSONWP Proxy] Proxying [POST /element] to [POST http://localhost:8200/wd/hub/session/e6f1a2d6-9ff2-4d95-aa5c-b403aa417150/element] with body: {"strategy":"id","selector":"com.ocito.laredoute:id/tv_selected_country","context":"","multiple":false}
[debug] [JSONWP Proxy] Got response with status 200: {"sessionId":"e6f1a2d6-9ff2-4d95-aa5c-b403aa417150","status":0,"value":{"ELEMENT":"166b0f24-8864-4c1c-b181-1eb32b38d8b4"}}
[debug] [MJSONWP] Responding to client with driver.findElement() result: {"ELEMENT":"166b0f24-8864-4c1c-b181-1eb32b38d8b4"}
[HTTP] <-- POST /wd/hub/session/a0ac2a7d-a5aa-428a-9044-c98d3061bbde/element 200 152 ms - 122
[HTTP] --> GET /wd/hub/session/a0ac2a7d-a5aa-428a-9044-c98d3061bbde/element/166b0f24-8864-4c1c-b181-1eb32b38d8b4/displayed {}
[debug] [MJSONWP] Calling AppiumDriver.elementDisplayed() with args: ["166b0f24-8864-4c1c-b181-1eb32b38d8b4","a0ac2a7d-a5aa-428a-9044-c98d3061bbde"]
[debug] [JSONWP Proxy] Proxying [GET /element/166b0f24-8864-4c1c-b181-1eb32b38d8b4/attribute/displayed] to [GET http://localhost:8200/wd/hub/session/e6f1a2d6-9ff2-4d95-aa5c-b403aa417150/element/166b0f24-8864-4c1c-b181-1eb32b38d8b4/attribute/displayed] with body: {}
[debug] [JSONWP Proxy] Got response with status 200: "{\"sessionId\":\"e6f1a2d6-9ff2-4d95-aa5c-b403aa417150\",\"status\":0,\"value\":\"true\"}"
[debug] [MJSONWP] Responding to client with driver.elementDisplayed() result: true
[HTTP] <-- GET /wd/hub/session/a0ac2a7d-a5aa-428a-9044-c98d3061bbde/element/166b0f24-8864-4c1c-b181-1eb32b38d8b4/displayed 200 88 ms - 76
[HTTP] --> GET /wd/hub/session/a0ac2a7d-a5aa-428a-9044-c98d3061bbde/element/166b0f24-8864-4c1c-b181-1eb32b38d8b4/name {}
[MJSONWP] Driver proxy active, passing request on via HTTP proxy
[debug] [JSONWP Proxy] Proxying [GET /wd/hub/session/a0ac2a7d-a5aa-428a-9044-c98d3061bbde/element/166b0f24-8864-4c1c-b181-1eb32b38d8b4/name] to [GET http://localhost:8200/wd/hub/session/e6f1a2d6-9ff2-4d95-aa5c-b403aa417150/element/166b0f24-8864-4c1c-b181-1eb32b38d8b4/nam
e] with body: {}
[debug] [JSONWP Proxy] Got response with status 200: "{\"sessionId\":\"e6f1a2d6-9ff2-4d95-aa5c-b403aa417150\",\"status\":0}"
[JSONWP Proxy] Replacing sessionId e6f1a2d6-9ff2-4d95-aa5c-b403aa417150 with a0ac2a7d-a5aa-428a-9044-c98d3061bbde
[HTTP] <-- GET /wd/hub/session/a0ac2a7d-a5aa-428a-9044-c98d3061bbde/element/166b0f24-8864-4c1c-b181-1eb32b38d8b4/name 200 44 ms - 63
[HTTP] --> GET /wd/hub/session/a0ac2a7d-a5aa-428a-9044-c98d3061bbde/screenshot {}
[debug] [MJSONWP] Calling AppiumDriver.getScreenshot() with args: ["a0ac2a7d-a5aa-428a-9044-c98d3061bbde"]
[debug] [ADB] Device API level: 23
[AndroidDriver] Cannot get screenshot data with 'adb exec-out' because of 'Command output is not a valid PNG file (fd504e47 !== 89504e47)'. Defaulting to 'adb shell' call
[debug] [ADB] Getting connected devices...
[debug] [ADB] 1 device(s) connected
[debug] [ADB] Running 'C:\Users\it140258\AppData\Local\Android\android-sdk\platform-tools\adb.exe' with args: ["-P",5037,"-s","3300ec51d0d3a269","shell","/system/bin/rm","/data/local/tmp/screenshot.png;","/system/bin/screencap","-p","/data/local/tmp/screenshot.png"]
[debug] [ADB] Running 'C:\Users\it140258\AppData\Local\Android\android-sdk\platform-tools\adb.exe' with args: ["-P",5037,"-s","3300ec51d0d3a269","pull","/data/local/tmp/screenshot.png","C:\\Users\\it140258\\AppData\\Local\\Temp\\appium117424-13472-om08f0.n6po4gf1or.png"]
[debug] [MJSONWP] Responding to client with driver.getScreenshot() result: "iVBORw0KGgoAAAANSUhEUgAABgAAAAgACAYAAADg0fg2AACAAElEQVR4AezBe2yd953n9/fnOYeHN5EiRSs2qUimLVuyrCi2RTsaW46iSdCkmcSDSYOi29nNZDtdtGi3i8UuEOxOWhTbNot2B9Mb2gEWLbZNsJiZ3T+2/cNuZ6ysN5rmIo1jQ5IZW5HGUugbSZs
6pkmKInnI83yqHzWPckSRFKmLFdnf10vf+ta3TAghhBBCCCGEEEIIIYQQPlIyQgghhBBCCCGEEEIIIYTwkZMRQgghhBBCCCGEEEIIIYSPnIwQQgghhBBCCCGEEEIIIXzkZIQQQgghhBBCCCGEEEII4SMnI4QQQgghhBBCCCGEEEIIHzkZIYQQQgghhBBCCCGEEEL4yMkIIYQQQgghhBBCCCGEEMJHTkYIIYQQQgghhBBCCCGEED5yyoQQQgghhBBCCCGEEMISlUqF/v
5+enp66OnpIbnrrrtIzp07R1KtVqlWqwwNDVGr1Qh3tq9+9av09PSwku9973uEO0uZEEIIIYQQQgghhBBCuKhSqbBjxw527txJT08PK+nr6yPp6+ujUK1WOXXqFKdPn6ZWqxHuPD09PWzevJnw0VEmhBBCCCGEEEIIIYTwsTcwMMCePXuoVCpcj56eHp566ikef/xxBgcHefnllwkh3F5lQgghhBBCCCGEEEIIH1u9vb0cPHiQjo4Oljp//jzDw8NMTU1RrVaZm5sja
W5upqenh46ODvr6+tiwYQOFSqXCwMAAO3bs4NChQ1SrVcKvlq9+9av09PSwVFdXF6v55je/yXK+973vEX41lQkhhBBCCCGEEEIIIXws7dmzhyeffJJGtVqN06dPc+rUKarVKisZGhqi0NPTw86dO9mxYweVSoWko6ODr3/96/zkJz/hZz/7GeFXR09PD5s3b2a9Nm/eTLizlAkhhBBCCCGEEEIIIXzs7NmzhyeffJJGp0+f5uWXX2ZqaopGlUqFnp4eGlWrVWq1Gkm1
WuUnP/kJg4OD...
[HTTP] <-- GET /wd/hub/session/a0ac2a7d-a5aa-428a-9044-c98d3061bbde/screenshot 200 5966 ms - 1486562
[HTTP] --> DELETE /wd/hub/session/a0ac2a7d-a5aa-428a-9044-c98d3061bbde {}
[debug] [MJSONWP] Calling AppiumDriver.deleteSession() with args: ["a0ac2a7d-a5aa-428a-9044-c98d3061bbde"]
[debug] [BaseDriver] Event 'quitSessionRequested' logged at 1495640787982 (16:46:27 GMT+0100 (Hora de Verão de GMT))
[debug] [UiAutomator2] Deleting UiAutomator2 session
[debug] [UiAutomator2] Deleting UiAutomator2 server session
[debug] [JSONWP Proxy] Proxying [DELETE /] to [DELETE http://localhost:8200/wd/hub/session/e6f1a2d6-9ff2-4d95-aa5c-b403aa417150] with no body
[debug] [JSONWP Proxy] Got response with status 200: "{\"sessionId\":\"e6f1a2d6-9ff2-4d95-aa5c-b403aa417150\",\"status\":0,\"value\":\"Session deleted\"}"
[debug] [ADB] Getting connected devices...
[debug] [ADB] 1 device(s) connected
[debug] [ADB] Running 'C:\Users\it140258\AppData\Local\Android\android-sdk\platform-tools\adb.exe' with args: ["-P",5037,"-s","3300ec51d0d3a269","shell","am","force-stop","com.ocito.laredoute"]
[debug] [UiAutomator2] FULL_RESET set to 'true', Uninstalling 'com.ocito.laredoute'
[debug] [ADB] Uninstalling com.ocito.laredoute
[debug] [ADB] Getting connected devices...
[debug] [ADB] 1 device(s) connected
[debug] [ADB] Running 'C:\Users\it140258\AppData\Local\Android\android-sdk\platform-tools\adb.exe' with args: ["-P",5037,"-s","3300ec51d0d3a269","shell","am","force-stop","com.ocito.laredoute"]
[debug] [ADB] Running 'C:\Users\it140258\AppData\Local\Android\android-sdk\platform-tools\adb.exe' with args: ["-P",5037,"-s","3300ec51d0d3a269","uninstall","com.ocito.laredoute"]
[ADB] App was uninstalled
[debug] [Logcat] Stopping logcat capture
[debug] [ADB] Removing forwarded port socket connection: 8200
[debug] [ADB] Running 'C:\Users\it140258\AppData\Local\Android\android-sdk\platform-tools\adb.exe' with args: ["-P",5037,"-s","3300ec51d0d3a269","forward","--remove","tcp:8200"]
[Appium] Removing session a0ac2a7d-a5aa-428a-9044-c98d3061bbde from our master session list
[debug] [BaseDriver] Event 'quitSessionFinished' logged at 1495640791562 (16:46:31 GMT+0100 (Hora de Verão de GMT))
[debug] [MJSONWP] Received response: null
[debug] [MJSONWP] But deleting session, so not returning
[debug] [MJSONWP] Responding to client with driver.deleteSession() result: null
[HTTP] <-- DELETE /wd/hub/session/a0ac2a7d-a5aa-428a-9044-c98d3061bbde 200 3585 ms - 76

I was wondering if something can be done on Cerberus side.

Note that this issue is not related to the latest deploy

Thanks and Regards
Cerberus Version : 1.1.13-SNAPSHOT-1150

**Comment by j0hnsa** [comment]
Here's a little explanation of Uiautomator2 driver comparing to Appium bootstrap:
https://github.com/appium/appium-uiautomator2-server/wiki/bootstrap-vs-appium-uiautomator2-server-module

Steps to replicate the problem i'm facing:

install appium from sources with the use of npm (npm install -g appium@latest and can be done also with --no-shrinkwrap the output will be the same ) that will install 1.6.5 version (stable) the configurations that I use on the robot can be checked below: 
![2017-06-06 13_50_29-robot](https://cloud.githubusercontent.com/assets/16339891/26829947/2e8ff4aa-4abf-11e7-87e3-f137c45495a5.png)

This ClassCastException will be reached on choose country page.

All I could find related to this exception seems to be related to Selenium:
https://github.com/seleniumhq/selenium-google-code-issue-archive/issues/5154

Can you please take a moment to test and analyse how can we overcome this issue? I really believe that the use of Uiautomator2 driver could fix most of the problems faced with the other automation method (pages stalling, slowness, badly type on some textboxes among others)

Thanks and Regards

**Comment by j0hnsa** [comment]
Setting used on Mac:
install with npm appium@latest (1.6.5)
I'm launching the appium server with the flag for a default capabilities file which looks like this:
![2017-06-06 15_02_36-mac la s mac mini - vnc viewer](https://cloud.githubusercontent.com/assets/16339891/26833104/65981f40-4ac9-11e7-92ce-d7b1ca6b76a2.png)

appium -dc PATH_TO_FILE -a YOUR_IP -p PORT

And the robot:

![2017-06-06 15_03_37-robot](https://cloud.githubusercontent.com/assets/16339891/26833120/723c5432-4ac9-11e7-85bc-835932cbc4d1.png)

Maybe not all the capabilities are mandatory for a good functioning of appium but with this settings it is quite stable so i didn't changed them.

**Comment by abourdon** [comment]
Hi @j0hnsa,

I'm able to execute test with UIAutomator2 by installing the 1.6.5 following the Cerberus documentation.

Bellow the configuration I used:

Appium version: 1.6.5
Java version: 1.8.0_112
Adb version: 1.0.32
OS: Mac OS Sierra
Robot configuration:
<img width="568" alt="screen shot 2017-06-07 at 11 31 47" src="https://user-images.githubusercontent.com/1511715/26871896-02029db4-4b75-11e7-867c-eff30565e083.png">

And then:

<img width="1474" alt="screen shot 2017-06-07 at 11 31 26" src="https://user-images.githubusercontent.com/1511715/26871957-3895be10-4b75-11e7-8c56-b6875cc0d413.png">

Could you please try to use same configuration (if you can) and try again?
Also, beware to correctly install Appium though the Cerberus documentation. Be sure to clean any build files as the node_modules folder.

Regards,
Aurélien

**Comment by j0hnsa** [comment]
Hello @abourdon ,
My settings aren't much different :

Java(TM) SE Runtime Environment (build 1.8.0_112-b16)
Android Debug Bridge version 1.0.39
Revision 5943271ace17-android
Appium 1.6.5

can you please try to run this test for instance:
http://192.168.134.17/Cerberus/ExecutionDetail2.jsp?executionId=47288054

also are you using the latest apk file? or should I send you a copy? as this error wasn't always present.

Regards

**Comment by bcivel** [comment]
Hi @j0hnsa 
Testcase you provided seems to have been corrected. 
Could you duplicate it and change the action again, and execute again for me to analyse your issue?
Thanks
Benoit

**Comment by j0hnsa** [comment]
Hello @bcivel ,

This error actually is not occurring anymore since I started using just the appium that we installed previously on the Mac to launch iOS tests (installed following cerberus documentation). weirdly if I follow the documentation now to install a more recent build of appium this issue will be present again, so I'm not sure if some step may be missing there.
Maybe if someone could just confirm that the steps to install and run appium is something like this for me to be able to run appium locally:

git clone appium
cd on to appium 
npm install -g mocha
npm install -g gulp
npm install -g gulp-cli
npm install
gulp transpile
node .

 
Now the only problem I'm finding when using Uiautomator2 is the one mentioned on issue #1347 that I also think that may be selenium related.

Regards
